### PR TITLE
Integrate new prompt mechanism into training

### PR DESCRIPTION
**Pull request by jeffreymeetkai**
- move train.py to root folder
- implement token and embedding resizing to add the new tokens
- fix compute_metrics to use `len(tokenizer)` instead of `tokenizer.vocab_size` when reshaping the tensors/arrays
- integrate prepare_training_inputs into CustomDataset class

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -3,18 +3,19 @@
...
 import torch.distributed
-from llama_flash_attn_monkey_patch import replace_llama_attn_with_flash_attn
+import transformers
 from torch.nn import CrossEntropyLoss
+from transformers import LlamaTokenizer, Trainer
 
-replace_llama_attn_with_flash_attn()
+from functionary.prompt import EndToken
+from train.custom_datasets import CustomDataset, split_data
+from train.llama_flash_attn_monkey_patch import replace_llama_attn_with_flash_attn
```
Comment (musab-mk) [inline train.py:16]: Transformers must be imported after the patch. Lets revert back the previous order

File: train.py [hunk 2]
```diff
@@ -88,6 +116,11 @@ def train():
         # legacy=False,  # See: https://github.com/huggingface/transformers/pull/24565
     )
     tokenizer.pad_token = tokenizer.unk_token
+    added_tokens = [e.value for e in EndToken]
+    special_tokens_dict = {"additional_special_tokens": added_tokens}
+    smart_tokenizer_and_embedding_resize(
+        special_tokens_dict=special_tokens_dict, tokenizer=tokenizer, model=model
+    )
```
Comment (musab-mk) [inline train.py:123]: I think we should refactor tokenizer initialization into seperate function

File: train/custom_datasets.py [hunk 3]
```diff
@@ -29,172 +37,86 @@ def split_data(raw_data, input_file, percentage):
...
-        return_tensors="pt",
-        padding="max_length",
-        truncation=True,
-        max_length=tokenizer.model_max_length,
+    # a dictionary mapping from token_id --> end_token
+    id_to_endtoken = get_token_id_to_end_token(tokenizer)
+    prompt_str = (
+        "system:\n"
+        + generate_schema_from_functions(functions=messages["functions"])
+        + "\nsystem:\nA chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. The assistant calls functions with appropriate input when necessary\n"
```
Comment (musab-mk) [inline train/custom_datasets.py:70]: We should let the get_prompt_from_messages function to prepare the input. So, prepend this as a message dict to the original messages.

Comment (musab-mk) [reply train/custom_datasets.py:70]: with proper role and content etc

File: train/custom_datasets.py [hunk 4]
```diff
@@ -29,172 +37,86 @@ def split_data(raw_data, input_file, percentage):
...
-        targets = create_target_tensors(input_ids)
-
-    return text, input_ids, targets
-
-
-def prepare_messages_for_model(messages, tokenizer):
-    """Prepares a list of messages for the model by calling `prepare_message_for_model` function on each of them and
-    concatenating the returned input_ids and targets. Also, the function merges the text of the messages.
+def prepare_training_inputs(
+    messages: List[Dict],
```
Comment (musab-mk) [inline train/custom_datasets.py:41]: are you sure this is a list?

Comment (musab-mk) [reply train/custom_datasets.py:41]: Maybe we should add some CI action to catch these kind of stuff

File: train/custom_datasets.py [hunk 5]
```diff
@@ -213,11 +135,11 @@ def __getitem__(self, i) -> Dict[str, torch.Tensor]:
         if i in self.cached_data_dict:
             return self.cached_data_dict[i]
 
-        ret = prepare_messages_for_model(self.raw_data[i], self.tokenizer)
+        ret = prepare_training_inputs(self.raw_data[i], self.tokenizer)
         ret = {
-            "input_ids": ret["input_ids"],
-            "labels": ret["labels"],
-            "attention_mask": ret["attention_mask"],
+            "input_ids": ret[1]["input_ids"],
```
Comment (musab-mk) [inline train/custom_datasets.py:140]: Why is this ret[1] ? Can we find a more explicit way. If we cannot, we should at least add some clarification

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: functionary/train/custom_datasets.py
Comment (musab-mk) [inline functionary/train/custom_datasets.py:1]: Can we rename this file into something much more appropriate

Comment (musab-mk) [reply functionary/train/custom_datasets.py:1]: In general, let's spend more time on naming things

File: functionary/prompt.py [hunk 6]
```diff
@@ -5,6 +5,10 @@
 
 import torch
 
+from functionary.schema import generate_schema_from_functions
+
+SYSTEM_MESSAGE = """A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. The assistant calls functions with appropriate input when necessary"""
```
Comment (musab-mk) [inline functionary/prompt.py:10]: There was already a system message in the inference. Either let's delete this and import that one. Or delete the other one.

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: tests/test_prompt_creation.py [hunk 7]
```diff
@@ -104,12 +161,16 @@ def test_correct_end_token(self):
...
+            self.assertFalse(
+                prompt.endswith(stop_token), f"`{prompt}` ends with: `{stop_token}`"
+            )
 
     def test_prepare_training_inputs(self):
         """this function is used to test function: prepare_training_inputs"""
         # note that must set legacy=True, read more: https://github.com/huggingface/transformers/issues/25176
-        tokenizer = LlamaTokenizer.from_pretrained("musabgultekin/functionary-7b-v1", legacy=True)
+        tokenizer = LlamaTokenizer.from_pretrained(
+            "musabgultekin/functionary-7b-v1", legacy=True
```
[the remark this answers is not in the document]
Comment (jeffreymeetkai) [reply tests/test_prompt_creation.py:172]: I've set both to using fast tokenizer with legacy=True.

File: tests/test_prompt_creation.py [hunk 8]
```diff
@@ -40,60 +36,121 @@ def extract_unmasked_chunks(labels: List[int]) -> List[List[int]]:
...
 
     def test_final_prompt_generation(self):
-        final_prompt = get_prompt_from_messages(self.test_case)
+        final_prompt = get_prompt_from_messages(
+            self.test_case["messages"], self.test_case["functions"]
+        )
+
         self.assertEqual(
-            final_prompt.strip(), self.final_prompt.strip(), "wrong final prompt from: get_prompt_from_messages"
+            final_prompt.strip(),
```
[the remark this answers is not in the document]
Comment (jeffreymeetkai) [reply tests/test_prompt_creation.py]: Fixed

**Review by musab-mk — APPROVED** [review]

**Pull request merged**